src/PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.cpp

```diff
@@ -196,19 +205,321 @@ Scalar<DataType> time_deriv_of_lapse(
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  tnsr::I<DataType, SpatialDim, Frame> dt_shift{};
+  GeneralizedHarmonic::time_deriv_of_shift<SpatialDim, Frame, DataType>(
+      make_not_null(&dt_shift), lapse, shift, inverse_spatial_metric,
+      spacetime_unit_normal, phi, pi);
+  return dt_shift;
+}
+
+namespace {
+template <size_t SpatialDim, typename Frame, typename DataType>
+struct D0LowerShiftBuffer;
+
+template <size_t SpatialDim, typename Frame>
+struct D0LowerShiftBuffer<SpatialDim, Frame, double> {
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  explicit D0LowerShiftBuffer(const size_t /*size*/) noexcept {}
+
+  tnsr::II<double, SpatialDim, Frame> inverse_spatial_metric;
+  tnsr::I<double, SpatialDim, Frame> dt_shift;
+  tnsr::ii<double, SpatialDim, Frame> dt_spatial_metric;
+};
+
+template <size_t SpatialDim, typename Frame>
+struct D0LowerShiftBuffer<SpatialDim, Frame, DataVector> {
+ private:
+  // We make one giant allocation so that we don't thrash the heap.
+  Variables<tmpl::list<::Tags::TempII<0, SpatialDim, Frame, DataVector>,
+                       ::Tags::TempI<1, SpatialDim, Frame, DataVector>,
+                       ::Tags::Tempii<2, SpatialDim, Frame, DataVector>>>
+      buffer_;
+ public:
+  explicit D0LowerShiftBuffer(const size_t size) noexcept
+      : buffer_(size),
+        inverse_spatial_metric(
+            get<::Tags::TempII<0, SpatialDim, Frame, DataVector>>(buffer_)),
+        dt_shift(get<::Tags::TempI<1, SpatialDim, Frame, DataVector>>(buffer_)),
+        dt_spatial_metric(
+            get<::Tags::Tempii<2, SpatialDim, Frame, DataVector>>(buffer_)) {}
+
+  tnsr::II<DataVector, SpatialDim, Frame>& inverse_spatial_metric;
+  tnsr::I<DataVector, SpatialDim, Frame>& dt_shift;
+  tnsr::ii<DataVector, SpatialDim, Frame>& dt_spatial_metric;
+};
+}  // namespace
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void time_deriv_of_lower_shift(
+    const gsl::not_null<tnsr::i<DataType, SpatialDim, Frame>*> dt_lower_shift,
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::ii<DataType, SpatialDim, Frame>& spatial_metric,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi) noexcept {
+  if (UNLIKELY(get_size(get<0>(*dt_lower_shift)) != get_size(get(lapse)))) {
+    *dt_lower_shift = tnsr::i<DataType, SpatialDim, Frame>(get(lapse));
+  }
+  // Use a Variables to reduce total number of allocations. This is especially
+  // important in a multithreaded environment.
+  D0LowerShiftBuffer<SpatialDim, Frame, DataType> buffer(get_size(get(lapse)));
```

Comment:
The use of this buffer is not helping, and is probably making things worse.  Two of the buffer entries are being initialized by copy from objects that immediately go out of scope, so we are still incurring allocations for the initializing objects.  If we just used local variables we could move into them (in one case implicitly, in the other I think an explicit `std::move` is required), so we still wouldn't need an allocation and could avoid the copy into the buffer.